lib/SILOptimizer/SILCombiner/SILCombinerMiscVisitors.cpp

```diff
@@ -1074,6 +1074,10 @@ SILInstruction *SILCombiner::visitReleaseValueInst(ReleaseValueInst *RVI) {
   SILValue Operand = RVI->getOperand();
   SILType OperandTy = Operand->getType();
 
+  // Do not remove a release that calls a value deinit.
+  if (hasValueDeinit(OperandTy))
```

Comment:
This check also needs to be done in `isTrivial` above. If the operand type is a struct without a deinit, but one of the struct fields has a deinit, the release_value will be removed. E.g.
```
  struct WithDeinit : ~Copyable {
    deinit {}
  }
  struct Outer {
    var wd: WithDeinit
  }

  ...
  release_value %1 : $Outer
```

Reply:
Right! That issue was fixed here, but you didn't see it because I rebased this PR on top of Kavon's
```
commit e20243219042458ff256bb6f88e53cc086cb3736 (public/5.9-fix-silcombine)
Author: Andrew Trick <atrick@apple.com>
Date:   Thu May 25 16:14:16 2023

    Fix SILCombine to preserve a release of a move-only type

    The release needs to be preserved in case a user-defined deinit is
    present in the released type. Checking for move-only is slightly
    conservative.

    Fixes rdar://109846094 ([move-only] SILCombine eliminates struct deinitialization)

    (cherry picked from commit dc974e8c00d923474905d9c41d6b71bc045751d9)
```